index.rb

```diff
@@ -0,0 +1,120 @@
+module Enumerable
+  def my_each
+    return enum_for unless block_given?
+
+    list = is_a?(Range) ? to_a : self
+    i = 0
+    while i < list.length
+      yield(list[i])
+      i += 1
+    end
+    list
+  end
+
+  def my_each_with_index()
+    return enum_for unless block_given?
+
+    list = is_a?(Range) ? to_a : self
+    index = 0
+    while index < list.length
+      yield(self[index], index)
+      index += 1
+    end
+  end
+
+  def my_select
+    return enum_for unless block_given?
+
+    filter = []
+    my_each { |element| filter.push(element) if yield(element) }
+    filter
+  end
+
+  def my_all?(*args)
+    lever = true
+    if !args[0].nil?
+      my_each { |element| lever = false unless args[0] == element }
+    elsif !block_given?
+      my_each { |element| lever = false unless element }
+    else
+      my_each { |element| lever = false unless yield(element) }
+    end
+    lever
+  end
+
+  def my_any?(*arg)
+    lever = false
+    if !arg[0].nil?
+      my_each { |element| lever = true if arg[0] == element }
+    elsif !block_given?
+      my_each { |element| lever = true if element }
+    else
+      my_each { |element| lever = true if yield(element) }
+    end
+    lever
+  end
+
+  def my_none?(arg = nil, &block)
+    !my_any?(arg, &block)
+  end
+
+  def my_count(num = nil)
+    count = 0
+    if num
+      my_each { |element| count += 1 if element == num }
+    elsif !block_given?
+      count = length
+    elsif !num
+      my_each { |element| count += 1 if yield element }
+    end
+    count
+  end
+
+  def my_map(*procs)
+    result = []
+    if procs.count.zero?
+      my_each do |elem|
+        result << yield(elem)
+      end
+    else
+      proc = procs[0]
+      my_each(&proc)
+    end
+    result
+  end
```

Comment:
* [ ] should have working `my_map` identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

#### my_map 

- [ ] returns an Enumerator if no block is given
```ruby
[2,4,7,11].my_map should be equal to [2,4,7,11].map #<Enumerator: [2, 4, 7, 11]:map
```
- Yours is retrieving a LocalJumpError
![image](https://user-images.githubusercontent.com/52209067/79290627-e244cb00-7e89-11ea-9bf8-421dfdc162f7.png)

- Real map
![image](https://user-images.githubusercontent.com/52209067/77697646-dc3b8880-6f74-11ea-9277-4ed4a2b3af4c.png)

[OPTIONAL]

- [ ] in the specifications of the project in the [Odin project](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks) # 12, you were asked to modify your my_map method to take either a proc or a block. And if both a proc and a block are given, only execute the proc. The real map doesn't have this feature, but you were asked to have it so I strongly recommend working on it

Given 
```ruby
my_proc = Proc.new {|num| num > 10 }
```
Result
```ruby
puts [18, 22, 5, 6] .my_map(my_proc) {|num| num < 10 }  => true true false false
```